text/0096-auto-batching.md

```diff
@@ -0,0 +1,67 @@
+- Title: Auto-Batching
+- Start Date: 2021-10-06
+- Specification PR: [#96](https://github.com/meilisearch/specifications/pull/96)
+- Discovery Issue: [#69](https://github.com/meilisearch/product/issues/69)
+
+# Auto-Batching
+
+## 1. Functional Specification
+
+### 1.1 Summary
+
+MeiliSearch can automatically group consecutive asynchronous `documentAddition` or `documentPartial` tasks for the same index into a single batch via an automatic batching mechanism.
+
+#### Summary Key points
+
+- A `batchUid` field is added on a fully qualified `task` API object.
+- An auto-batching mechanism groups consecutive `documentAddition` or consecutive `documentPartial` enqueued tasks for a similar index when a task is fetched from the FIFO task queue to be processed.
+
+### 1.2 Motivation
+
+We regularly tell users to batch their documents to speed up the indexing speed. We decided to integrate a simple scheduler to automatically batch consecutive tasks for an index to make it transparent for the users and enhance their quality of life using MeiliSearch.
+
+### 1.3 Explanations
+
+#### 1.3.1 Overall
+
+All `tasks` are linked to a specific `batchUid` field. This `batchUid` field permits to identify that several identical, consecutive tasks have been grouped in the same batch.
+
+Only consecutive `documentAddition` and `documentPartial` `tasks` for the same index can have a shared `batchUid` since the scheduler will be able to group them together. This means that all `tasks` concerning other operations will also be part of a `batchUid` having only one task. This could be extended in the future by enhancing the scheduler capabilities.
+
+This is to avoid having multiple incompressible computation time by grouping them together to increase performance when applied to a maximum number of documents.
+
+#### 1.3.2 Grouping tasks to a single batch
+
+The scheduling program that groups tasks within a single batch is triggered when an asynchronous `task` currently processed reaches a terminal state as `succeeded` or `failed`.
+
+In other words, when the next `task` should be picked from the FIFO task queue. The scheduler fetches and groups all consecutive `documentAddition` for a similar index in a batch until it encounters another task type or a similar task type but for a different index.
+
+> Note that we are considering implementing a configurable limit of maximum documents to process for a batch and a flag to activate the auto-batching mechanism.
```

Comment:
So basically a way of telling an instance "batch `documentAddition` up to X amount of documents" so we avoid memory consumption problems?